[CS-4727] Logic and flow for push notifications permissions screen

oidoug opened the pull request:
### Description

This PR adds logic to show or skip the Notification Permissions Screen. After skipped once the screen will not show again, but users can go to notifications in Settings to be presented with the notification permission again if still pending. Skipping is important to avoid spamming users that don't want notifications at all.

The screen was added to the onboarding flow and before backup, as first intended for the onboarding flow.

- [x] Completes #(CS-4727)

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

### Screenshots

https://user-images.githubusercontent.com/129619/206207559-9da34f5b-5ddf-4f91-a448-096201fcc253.mp4

oidoug committed:
[CS-4727] Implements logic for useNotificationsPermissionScreen, including tests;
Simplifies param types for onUpdateOptionStatus call on useUpdateNotificationPreferences hook;
Includes new screen in wallet initialization flow if needed;

oidoug committed:
[CS-4727] Adds new persistent flag for Notif. Perm. skip user selection;
[CS-4727] Changes unchecked checkbox color to be slightly darker/transparent;
[CS-4727] Moves Notif. Perm. flow logic out of wallet manager;
[CS-4727] Updates useShowOnboarding to acomodate new flow, with added tests.

oidoug committed:
[CS-4727] Fixes broken tests on useShowOnboarding and useNotificationsPermissionsScreen.

danibonilha commented on cardstack/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -32,11 +32,11 @@ export const Checkbox = ({
   const [disabled] = useState(isDisabled);
```
I know it's not the purpose of this PR, nor it was you who wrote this, but I think we can remove this state, and use the disabled property directly

danibonilha commented on cardstack/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -32,11 +32,11 @@ export const Checkbox = ({
+      onPress(!selected);
     }
+
+    setSelected(!selected);
```
nit: we could assign `!selected ` to a variable to make sure we use the same value

danibonilha commented on cardstack/src/models/firebase.ts:
```diff
@@ -203,7 +203,7 @@ export const checkPushPermissionAndRegisterToken = async () => {
             onPress: async () => {
-              resolve(true);
+              resolve(false);
             },
```
in this case, shouldn't the catch block return resolve as false too ?

oidoug replied:
Yes, that makes sense 👍

danibonilha commented on cardstack/src/screens/NotificationsPermissionScreen/NotificationsPermissionScreen.tsx:
```diff
@@ -24,40 +23,37 @@ const NotificationsPermissionScreen = () => {
     options,
     isError,
-    handleSkipOnPress,
+    handleOnUpdateOption,
```
supernit: I'd suggest `handleUpdateOption` or `onUpdateOption`

danibonilha commented on cardstack/src/screens/NotificationsPermissionScreen/NotificationsPermissionScreen.tsx:
```diff
@@ -24,40 +23,37 @@ const NotificationsPermissionScreen = () => {
-    handleSkipOnPress,
+    handleOnUpdateOption,
     handleEnableNotificationsOnPress,
+    handleSkipOnPress,
```
ditto: `handleSkipPress` or `onSkipPress`

danibonilha commented on cardstack/src/screens/NotificationsPermissionScreen/NotificationsPermissionScreen.tsx:
```diff
@@ -105,6 +101,8 @@ const NotificationsPermissionScreen = () => {
     []
   );
 
+  const keyExtractor = (item: NotificationsOptionsType) => item.type;
```
nit: could wrap on useCallback

danibonilha commented on cardstack/src/screens/NotificationsPermissionScreen/useNotificationsPermissionScreen.ts:
```diff
@@ -1,22 +1,49 @@
+    async (option: NotificationsOptionsType, isEnabled: boolean) =>
+      onUpdateOptionStatus(option.type, isEnabled),
+    [onUpdateOptionStatus]
+  );
```
do we really need this function at all? if we are just calling `onUpdateOptionStatus` with the same params, except it's option, vs option.type, but it seems we could just use onUpdateOptionStatus on the component

danibonilha replied:
On second though, shouldn't we wait for the user to hit enable, and give permission, before setting the notification options ? What's the point of storing this info if the user doesn't have an fcm token

oidoug replied:
I thought the same, but we don't actually use the fcm-token to save this info, just the hub auth token, and it seemed simpler to just save the user selection, than to somehow invalidate the selection if they skip, feels weird to not save the way the UI works. 
Not sure though.

danibonilha replied:
Got it, I know it doesn't use the fcmToken, but without it, it means the user never gets the notification, so we are storing related info on our db to never be used, I think it's ok for now giving the amount of users, but that might be cumbersome in the future, let's just keep that in mind, so then my first question takes place again, we may not need this handle function, it;s just wrapping the onUpdateOptions, if the intention is to "rename", we can just pass 
 `{  handleUpdateOption:  onUpdateOptionStatus }` as part of the hook return.

danibonilha commented on cardstack/src/models/firebase.ts:
```diff
@@ -203,7 +203,7 @@ export const checkPushPermissionAndRegisterToken = async () => {
+              resolve(false);
             },
             style: 'cancel',
             text: 'Dismiss',
```
also since we are messing with this function, could you replace the lang.t messages with simple strings ?

oidoug replied:
Maybe we don't need this confirmation Alert anymore... Because we already have the whole screen as a "push will be enabled" disclaimer. WDYT?

danibonilha replied:
Agreed for the initial flow, but maybe it's needed for the settings one

danibonilha commented on cardstack/src/types/NotificationsPreferenceDataType.ts:
```diff
@@ -1,6 +1,18 @@
+  description: string;
+  status: NotificationStatusType;
+}
+
```
👏

danibonilha commented on cardstack/src/hooks/onboarding/useShowOnboarding.ts:
```diff
@@ -46,6 +48,14 @@ export const useShowOnboarding = () => {
         return;
       }
 
+      const askForNotificationsPermissions = await needsToAskForNotificationsPermissions();
```
nit: maybe we could rename this function, this gives the impression that permission will be "asked"/requested, when in fact it just checking if permission was already granted. Maybe something like, `needsNotificationPermission`  or not negate the isAuthorized variable on the function, so then we can use something like `hasNotificationPermission` and negate it accordingly with usage

oidoug replied:
I prefer `needsNotificationPermission`, negating the response seems odd.

danibonilha commented on cardstack/src/hooks/onboarding/useShowOnboarding.ts:
```diff
@@ -54,9 +64,16 @@ export const useShowOnboarding = () => {
 
       if (shouldShowProfileCreationFlow) {
```
let's add a comment here explaining why we need this step, it's a bit confusing that `navigateToNextOnboardingStep` has steps as optionals, so it would be nice to add some sort of explanation on why this is needed

oidoug replied:
Yeah, it's confusing because we want to navigate based on an effect "sometimes" and some other times we need to navigate before the effect change takes place. I'm refactoring this function a bit and adding some comments, plz take a look to see if it's somewhat improved.

danibonilha commented on cardstack/src/screens/NotificationsPermissionScreen/useNotificationsPermissionScreen.ts:
```diff
@@ -1,22 +1,49 @@
+
+  const handleSkipOnPress = useCallback(() => {
+    triggerSkipNotificationPermission();
+    navigateToNextOnboardingStep();
```
Skipping shouldn't take the user to the profile creation ?

oidoug replied:
The second screen in the flow is the backup one. I think the flow is correct.

danibonilha commented on cardstack/src/screens/NotificationsPermissionScreen/useNotificationsPermissionScreen.ts:
```diff
@@ -1,22 +1,49 @@
+
+    navigateToNextOnboardingStep(Routes.BACKUP_EXPLANATION);
+  }, [triggerSkipNotificationPermission, navigateToNextOnboardingStep]);
+
```
it feels to me, that it doesn't matter if it's accepted or not, skipped or not, the next 'step' behavior can be the same, triggerSkipNotificationPermission then navigateToNextOnboardingStep(Routes.BACKUP_EXPLANATION); since we are not waiting to set the preferences after users permissions

oidoug replied:
Maybe theres some overlap but it's not really the same. If the user accepts (don't skip) and then later removes permissions, we'll ask again.

I changed how the navigation works a bit, see if it's any better.

danibonilha commented on cardstack/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -54,7 +54,13 @@ export const Checkbox = ({
+            ? 'underlineGray'
+            : selected
+            ? 'checkboxSelected'
+            : 'checkboxDeselected'
```
Doesn't this change the color on screens that we don't want too ? Or is it a new pattern for our checkbox ?  Also this nested ternary is a bit scary hehe, it seems we never use this disabled property, so maybe we can remove it ?

oidoug replied:
We need to differentiate the "fill" color somehow. I tested in white backgrounds and it's not optimal but better than using the same color for sure.

I removed the disabled as suggested.

oidoug committed:
[CS-4725] Removes unused isDisabled state from Checkbox;
[CS-4725] Resolves push check as false on exception;
[CS-4725] Refactor callback names.

oidoug committed:
[CS-4725] Renames notification permission check function

oidoug committed:
[CS-4725] Refactor onboarding navigate functions for better redability

oidoug committed:
[CS-4727] Updates navigation to Backup Explanation.

oidoug committed:
[CS-4727] Fixes test on useNotificationPermissionScreen.test

danibonilha commented on cardstack/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -22,21 +22,21 @@ const CHECKBOX_SIZE = 22;
 
   const handleCall = useCallback(() => {
-    setSelected(!selected);
+    const currentSelected = selected;
```
hehe, the suggestion was to make sure we use exact the same value with the negation, so 
`const currentSelected = !selected;` but then maybe the name should change,
 `const toggledCheck = !selected;`

oidoug replied:
Ah I see 👍

danibonilha commented on cardstack/src/hooks/onboarding/useShowOnboarding.ts:
```diff
@@ -33,33 +35,53 @@ export const useShowOnboarding = () => {
+
+  // Will navigate imperactively to provided route.
+  const navigateOnboardingTo = useCallback(
+    (route: string) => {
```
could we type it as route ?

oidoug replied:
Sure, do you mean `RouteNames` from global, right?

danibonilha replied:
Yes

oidoug committed:
[CS-4727] Adds handling for notif. permission DENIED status;
[CS-4727] Handles notif. permissions status better on Notifications Settings screen;
[CS-4727] Improved code readability.

oidoug committed:
[CS-4727] Removes unused import

oidoug committed:
[CS-4727] Better type for route.

oidoug committed:
[CS-4727] Simplifies code on firebase model, NotificationsPermission screen and hook, and on NotificationSection.

danibonilha commented on src/components/settings-menu/NotificationsSection.tsx:
```diff
@@ -1,12 +1,42 @@
+
+import { Alert } from '@rainbow-me/components/alerts';
+
+const buildAlert = (
```
since the alert is triggered here,  we could rename it to showAlert

danibonilha reviewed (approved):
🚀

oidoug committed:
[CS-4727] Renames buildAlert to showAlert
